src/compiler/checker.ts

```diff
@@ -6011,7 +6017,88 @@ namespace ts {
+                                    // Example:
+                                    //
+                                    //     let x: A | B | C;
+                                    //
+                                    //     if (isA(x)) { // isA(...) clause returns A and the ongoing narrowing type has
+                                    //                   // the value of A and it is a distinct type so narrow x to an
+                                    //                   // empty object type.
+                                    //
+                                    //     }
+                                    //     else if (isB(x)) { //  narrow x to A
+                                    //     }
+                                    //     else if (isC(x))  // narrow x to A | B
+                                    //     }
+                                    //     else { // x begins with A | B | C
+                                    //
+                                    //          x // an empty object type
+                                    //     }
+                                    //
+                                    if (!(type.flags & TypeFlags.Union)) {
+                                        if (originalType.flags & TypeFlags.Union &&
+                                            // Use the original type to check the narrowed type instead of the ongoing narrowing type.
+                                            // This is to deal with cases like below:
+                                            //
+                                            //     let x: A | B | C;
+                                            //
+                                            //     if (isB(x)) { // narrow x to A.
+                                            //     }
+                                            //     else if (isB(x)) { // narrow x to A
+                                            //     }
+                                            //     else if (isC(x)) { // narrow x to A | B
+                                            //     }
+                                            //     else { // x begins with A | B | C
+                                            //         x
+                                            //     }
+                                            //
+                                            // Which is the same as:
+                                            //
+                                            //     let x: A | B | C;
+                                            //
+                                            //     if (isA(x)) { // narrow x to A (but also removes A).
+                                            //     }
+                                            //     else if (isB(x)) { // narrow x to A
+                                            //     }
+                                            //     else if (isC(x)) {// narrow x to A | B

```

Comment:
I'm referring to `x` in the else clause below. So it narrows it to `A | B` by removing `C` from `A | B | C`. If I was referring to a `x` inside this `else if(sC(x))` then it will narrow to `C`.